From 37f732c9f98a6f476da905e6653ce7faaea6f78d Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Thu, 28 Aug 2014 15:03:04 +0900
Subject: [PATCH] API: BossBar API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ec58741..11d0ea7 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1086,6 +1086,31 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
         public void setRevertedHunger(boolean useOldHunger) {
             throw new UnsupportedOperationException("Not implemented yet.");
         }
+
+        /**
+         * Sets the boss bar message with value
+         * @param message the message player wish to have
+         * @param value the value between 1-200
+         */
+        public void setBarMessageWithValue(String message , float value){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+
+        /**
+         * Destroys the bar message, if they have any
+         */
+        public void destroyBarMessage(){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+
+        /**
+         * Sets the boss bar message with value
+         * @param message the message player wish to have
+         * @param value the value between 1-200
+         */
+        public void updateMessageWithValue(String message , float value){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
     }
 
     Kyoka kyoka();
-- 
1.9.4.msysgit.1

