From 0d56ff5935ff3ed36cbdfc193df5a86a5a0c8647 Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Wed, 12 Nov 2014 15:02:15 +0900
Subject: [PATCH] API: List-like Sidebar API


diff --git a/src/main/java/nl/hiraku/kyoka/Kyoka.java b/src/main/java/nl/hiraku/kyoka/Kyoka.java
index 55f51bc..bc12faf 100644
--- a/src/main/java/nl/hiraku/kyoka/Kyoka.java
+++ b/src/main/java/nl/hiraku/kyoka/Kyoka.java
@@ -3,4 +3,6 @@ package nl.hiraku.kyoka;
 public interface Kyoka {
 
     public BarManager getBarManager();
+
+    public boolean isEnhancedSidebarAvailable();
 }
diff --git a/src/main/java/nl/hiraku/kyoka/SideBar.java b/src/main/java/nl/hiraku/kyoka/SideBar.java
new file mode 100644
index 0000000..89d8277
--- /dev/null
+++ b/src/main/java/nl/hiraku/kyoka/SideBar.java
@@ -0,0 +1,22 @@
+package nl.hiraku.kyoka;
+
+public interface SideBar {
+
+    public void add(String element);
+
+    public void set(int index , String element);
+
+    public void remove(int index);
+
+    public void setName(String name);
+
+    public String getName();
+
+    public void clear();
+
+    public int size();
+
+    public void update();
+
+}
+
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ec58741..92352bb 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1086,6 +1086,13 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
         public void setRevertedHunger(boolean useOldHunger) {
             throw new UnsupportedOperationException("Not implemented yet.");
         }
+
+        /**
+         * Gets the Enhanced Sidebar , if available
+         */
+        public nl.hiraku.kyoka.SideBar getSideBar() {
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
     }
 
     Kyoka kyoka();
-- 
1.9.3 (Apple Git-50)

