From a855a76f9871daddc45846fe577afc5a38a66388 Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Sat, 30 Aug 2014 14:41:32 +0900
Subject: [PATCH] API: Ability to send its dead body to another player


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 11d0ea7..8eedeee 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -16,6 +16,7 @@ import org.bukkit.WeatherType;
 import org.bukkit.command.CommandSender;
 import org.bukkit.conversations.Conversable;
 import org.bukkit.map.MapView;
+import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
 import org.bukkit.scoreboard.Scoreboard;
 
@@ -1111,6 +1112,9 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
         public void updateMessageWithValue(String message , float value){
             throw new UnsupportedOperationException("Not implemented yet.");
         }
+        public void dropDeadBodyCopy(Player viewer , Plugin plugin){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
     }
 
     Kyoka kyoka();
-- 
1.9.4.msysgit.1

