From 4bd4c80eb1a3fbd6b0ebe003c48f400715356728 Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Sun, 31 Aug 2014 01:40:24 +0900
Subject: [PATCH] API: Allow Objective to be per Team for SideBar purposes


diff --git a/src/main/java/org/bukkit/scoreboard/Objective.java b/src/main/java/org/bukkit/scoreboard/Objective.java
index 321aac7..8699b20 100644
--- a/src/main/java/org/bukkit/scoreboard/Objective.java
+++ b/src/main/java/org/bukkit/scoreboard/Objective.java
@@ -107,4 +107,14 @@ public interface Objective {
      * @throws IllegalStateException if this objective has been unregistered
      */
     Score getScore(String entry) throws IllegalArgumentException, IllegalStateException;
+
+    public class Kyoka{
+        public void setTeamSpecific(boolean specific){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+        public boolean isTeamSpecific(){
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+    }
+    public Kyoka kyoka();
 }
-- 
1.9.4.msysgit.1

