From 992adf3fc7ccc6708cb513122eb57b49d9bb6ccc Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Tue, 11 Nov 2014 23:27:42 +0900
Subject: [PATCH] CORE: Non plugin referensing queue


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index e4dca37..8818885 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -598,6 +598,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
 
         // CraftBukkit start
         this.server.getScheduler().mainThreadHeartbeat(this.ticks);
+        nl.hiraku.kyoka.KyokaTicker.tick(this.ticks);
 
         // Run tasks that are waiting on processing
         while (!processQueue.isEmpty()) {
diff --git a/src/main/java/nl/hiraku/kyoka/KyokaTicker.java b/src/main/java/nl/hiraku/kyoka/KyokaTicker.java
new file mode 100644
index 0000000..9b1cf17
--- /dev/null
+++ b/src/main/java/nl/hiraku/kyoka/KyokaTicker.java
@@ -0,0 +1,40 @@
+package nl.hiraku.kyoka;
+
+import com.google.common.base.Preconditions;
+
+import java.util.*;
+
+public class KyokaTicker {
+
+    private static List<KyokaTimer> tickable = new ArrayList<>();
+
+    private static Map<Runnable , Integer> delayed = new HashMap<>();
+
+    public static void tick(int currentick){
+        Iterator<KyokaTimer> tickableIter = tickable.iterator();
+        while(tickableIter.hasNext()){
+            KyokaTimer timer = tickableIter.next();
+            if(!timer.tick()){
+                tickableIter.remove();
+            }
+        }
+        Iterator<Map.Entry<Runnable , Integer>> entryIterator = delayed.entrySet().iterator();
+        while(entryIterator.hasNext()){
+            Map.Entry<Runnable , Integer> entry = entryIterator.next();
+            entry.setValue(entry.getValue() - 1);
+            if(entry.getValue() <= 0){
+                entry.getKey().run();
+                entryIterator.remove();
+            }
+        }
+    }
+
+    static void delay(Runnable runnable , int tick){
+        delayed.put(runnable , tick );
+    }
+
+    static void addTimer(KyokaTimer tick){
+        Preconditions.checkState(!tickable.contains(tick));
+        tickable.add(tick);
+    }
+}
diff --git a/src/main/java/nl/hiraku/kyoka/KyokaTimer.java b/src/main/java/nl/hiraku/kyoka/KyokaTimer.java
new file mode 100644
index 0000000..b965a2c
--- /dev/null
+++ b/src/main/java/nl/hiraku/kyoka/KyokaTimer.java
@@ -0,0 +1,6 @@
+package nl.hiraku.kyoka;
+
+interface KyokaTimer {
+
+    public boolean tick();
+}
-- 
1.9.4.msysgit.2

