From bea5de70cf146edc819860eef3746d05fa159fcc Mon Sep 17 00:00:00 2001
From: Jamie Shaw <lk@lkay.org>
Date: Wed, 2 Jul 2014 20:49:37 +1000
Subject: [PATCH] API: Skeleton


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a0e27bc..0b2514a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1285,4 +1285,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         collection.add(new AttributeModifiable(getHandle().bb(), (new AttributeRanged("generic.maxHealth", scaledHealth ? healthScale : getMaxHealth(), 0.0D, Float.MAX_VALUE)).a("Max Health").a(true)));
     }
+
+    // Kyoka start
+    private final Kyoka kyoka = new Kyoka() {
+    };
+
+    public Kyoka kyoka() {
+        return kyoka;
+    }
+    // Kyoka end
 }
-- 
1.9.1

