From 89e851144c29b724448d6f24d5a667247f75d454 Mon Sep 17 00:00:00 2001
From: Jamie Shaw <lk@lkay.org>
Date: Wed, 2 Jul 2014 20:51:56 +1000
Subject: [PATCH] API: Allow access to player locale


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index c84b7b3..26ae01a 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -31,7 +31,7 @@ import org.spigotmc.ProtocolData; // Spigot - protocol patch
 public class EntityPlayer extends EntityHuman implements ICrafting {
 
     private static final Logger bL = LogManager.getLogger();
-    private String locale = "en_US";
+    public String locale = "en_US"; // Kyoka
     public PlayerConnection playerConnection;
     public final MinecraftServer server;
     public final PlayerInteractManager playerInteractManager;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a54a9d0..dc8b2ec 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1300,6 +1300,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     // Kyoka start
     private final Kyoka kyoka = new Kyoka() {
+        @Override
+        public String getLocale() {
+            return getHandle().locale;
+        }
     };
 
     public Kyoka kyoka() {
-- 
1.9.4.msysgit.2

