From e49426b783e6853dd8c917a7c0c9ae5a381bcfe4 Mon Sep 17 00:00:00 2001
From: Jamie Shaw <lk@lkay.org>
Date: Wed, 2 Jul 2014 21:41:22 +1000
Subject: [PATCH] FEAT: Fire BlockBreakEvent for extinguishing fire


diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 2db45ad..7c5fe71 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -105,6 +105,16 @@ public class PlayerInteractManager {
             if (event.isCancelled()) {
                 // Let the client know the block still exists
                 ((EntityPlayer) this.player).playerConnection.sendPacket(new PacketPlayOutBlockChange(i, j, k, this.world));
+
+                // Kyoka start - From SportBukkit!
+                org.bukkit.block.Block block = this.world.getWorld().getBlockAt(i, j, k).getRelative(org.bukkit.craftbukkit.block.CraftBlock.notchToBlockFace(l));
+                if (block == Blocks.FIRE) {
+                    i = block.getX();
+                    j = block.getY();
+                    k = block.getZ();
+                }
+                // Kyoka end
+
                 // Update any tile entity data for this block
                 TileEntity tileentity = this.world.getTileEntity(i, j, k);
                 if (tileentity != null) {
@@ -114,7 +124,7 @@ public class PlayerInteractManager {
             }
             // CraftBukkit end
             if (this.isCreative()) {
-                if (!this.world.douseFire((EntityHuman) null, i, j, k, l)) {
+                if (!this.world.douseFire(this.player, i, j, k, l)) { // Kyoka
                     this.breakBlock(i, j, k);
                 }
             } else {
@@ -137,7 +147,7 @@ public class PlayerInteractManager {
                     block.attack(this.world, i, j, k, this.player);
                     f = block.getDamage(this.player, this.player.world, i, j, k);
                     // Allow fire punching to be blocked
-                    this.world.douseFire((EntityHuman) null, i, j, k, l);
+                    this.world.douseFire(this.player, i, j, k, l); // Kyoka
                 }
 
                 if (event.useItemInHand() == Event.Result.DENY) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index eb5da2a..7148f1e 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1750,8 +1750,13 @@ public abstract class World implements IBlockAccess {
         }
 
         if (this.getType(i, j, k) == Blocks.FIRE) {
-            this.a(entityhuman, 1004, i, j, k, 0);
-            this.setAir(i, j, k);
+            // Kyoka start
+            if (entityhuman instanceof EntityPlayer) {
+                if (((EntityPlayer) entityhuman).playerInteractManager.breakBlock(i, j, k)) {
+                    this.triggerEffect(1004, i, j, k, 0);
+                }
+            }
+            // Kyoka end
             return true;
         } else {
             return false;
-- 
1.9.1

