From: Jamie Shaw <lk@lkay.org>
Date: Fri, 4 Jul 2014 18:00:39 +1000
From fa0316632167fefbbd90a99f1370b2929c094568 Mon Sep 17 00:00:00 2001
Subject: [PATCH] API: Entity collision


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 1d6886c..d68994d 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -432,7 +432,7 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
 
             List list = this.world.getEntities(this, axisalignedbb);
 
-            if (list != null) {
+            if (list != null && this.R()) { // Kyoka
                 for (int i = 0; i < list.size(); ++i) {
                     Entity entity = (Entity) list.get(i);
 
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index a44ad51..762c6d9 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1577,7 +1577,7 @@ public abstract class EntityLiving extends Entity {
     protected void bo() {
         List list = this.world.getEntities(this, this.boundingBox.grow(0.20000000298023224D, 0.0D, 0.20000000298023224D));
 
-        if (list != null && !list.isEmpty()) {
+        if (this.R() && list != null && !list.isEmpty()) { // Kyoka
             for (int i = 0; i < list.size(); ++i) {
                 Entity entity = (Entity) list.get(i);
 
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 26ae01a..e91e5a2 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -64,6 +64,19 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public double maxHealthCache;
     public boolean joining = true;
     // CraftBukkit end
+    // Kyoka start
+    public boolean collidesWithEntities = true;
+
+    @Override
+    public void Q() {
+        super.Q();
+    }
+
+    @Override
+    public boolean R() {
+        return this.collidesWithEntities && super.R();
+    }
+    // Kyoka end
 
     public EntityPlayer(MinecraftServer minecraftserver, WorldServer worldserver, GameProfile gameprofile, PlayerInteractManager playerinteractmanager) {
         super(worldserver, gameprofile);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index b22ca1b..07c666c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1363,6 +1363,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         public boolean canSee(Player player) {
             return !kyokaHiddenPlayers.contains(player);
         }
+
+        @Override
+        public boolean getCollidesWithEntities() {
+            return getHandle().collidesWithEntities;
+        }
+
+        @Override
+        public void setCollidesWithEntities(boolean shouldCollide) {
+            getHandle().collidesWithEntities = shouldCollide;
+            getHandle().k = shouldCollide;
+        }
     };
 
     public Kyoka kyoka() {
-- 
1.9.1

