From fb1d6ce80817de096240f6903858e4a94d080d13 Mon Sep 17 00:00:00 2001
From: Jamie Shaw <lk@lkay.org>
Date: Sun, 24 Aug 2014 16:59:28 +1000
Subject: [PATCH] FEAT: API for item despawn time and make it configurable


diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index c0cdd61..9dc6da9 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.List;
 
 // CraftBukkit start
+import nl.hiraku.kyoka.KyokaConfig;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent;
@@ -144,7 +145,7 @@ public class EntityArrow extends Entity implements IProjectile {
 
             if (block == this.g && i == this.h) {
                 ++this.at;
-                if (this.at == 1200) {
+                if (this.at >= KyokaConfig.arrowDespawnTicks) {
                     this.die();
                 }
             } else {
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 834def0..831915c 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 import java.util.Iterator;
 
+import nl.hiraku.kyoka.KyokaConfig;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -18,6 +19,7 @@ public class EntityItem extends Entity {
     public float c;
     private int lastTick = MinecraftServer.currentTick; // CraftBukkit
     public Entity owner; // Kyoka
+    public int despawnTicks = KyokaConfig.itemDespawnTicks; // Kyoka
 
     public EntityItem(World world, double d0, double d1, double d2) {
         super(world);
@@ -105,7 +107,7 @@ public class EntityItem extends Entity {
             }
 
             // ++this.age; // CraftBukkit - Moved up
-            if (!this.world.isStatic && this.age >= 6000) {
+            if (!this.world.isStatic && this.age >= despawnTicks) {
                 // CraftBukkit start - fire ItemDespawnEvent
                 if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
                     this.age = 0;
@@ -321,7 +323,7 @@ public class EntityItem extends Entity {
     public ItemStack getItemStack() {
         ItemStack itemstack = this.getDataWatcher().getItemStack(10);
 
-        return itemstack == null ? new ItemStack(Blocks.STONE) : itemstack; 
+        return itemstack == null ? new ItemStack(Blocks.STONE) : itemstack;
     }
 
     public void setItemStack(ItemStack itemstack) {
diff --git a/src/main/java/nl/hiraku/kyoka/KyokaConfig.java b/src/main/java/nl/hiraku/kyoka/KyokaConfig.java
index 3ebbc7c..f58d9e4 100644
--- a/src/main/java/nl/hiraku/kyoka/KyokaConfig.java
+++ b/src/main/java/nl/hiraku/kyoka/KyokaConfig.java
@@ -15,6 +15,8 @@ public class KyokaConfig {
     private static YamlConfiguration config;
 
     public static boolean bungeecord;
+    public static int itemDespawnTicks;
+    public static int arrowDespawnTicks;
 
     public static void init() {
         config = YamlConfiguration.loadConfiguration(configFile);
@@ -42,4 +44,14 @@ public class KyokaConfig {
         config.addDefault("bungeecord", false);
         bungeecord = config.getBoolean("bungeecord");
     }
+
+    private static void itemDespawnTicks() {
+        config.addDefault("item-despawn-ticks", 6000);
+        itemDespawnTicks = config.getInt("item-despawn-ticks");
+    }
+
+    private static void arrowDespawnTicks() {
+        config.addDefault("arrow-despawn-ticks", 1200);
+        arrowDespawnTicks = config.getInt("arrow-despawn-ticks");
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
index 02b32cf..4394cd6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
@@ -45,4 +45,16 @@ public class CraftItem extends CraftEntity implements Item {
     public EntityType getType() {
         return EntityType.DROPPED_ITEM;
     }
+
+    // Kyoka start
+    @Override
+    public int getDespawnTicks() {
+        return item.despawnTicks;
+    }
+
+    @Override
+    public void setDespawnTicks(int despawnTicks) {
+        item.despawnTicks = despawnTicks;
+    }
+    // Kyoka end
 }
-- 
1.9.1

