From cc99577eeddfae271f7520cec5e1fd72b85394bd Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Thu, 28 Aug 2014 16:17:25 +0900
Subject: [PATCH] API: BossBar-API


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
index 5829535..9a70017 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -4,18 +4,18 @@ import java.util.List;
 
 public class PacketPlayOutSpawnEntityLiving extends Packet {
 
-    private int a;
-    private int b;
-    private int c;
-    private int d;
-    private int e;
+    public int a;
+    public int b;
+    public int c;
+    public int d;
+    public int e;
     private int f;
     private int g;
     private int h;
     private byte i;
     private byte j;
     private byte k;
-    private DataWatcher l;
+    public DataWatcher l;
     private List m;
 
     public PacketPlayOutSpawnEntityLiving() {}
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a8b6ff5..171e119 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1381,6 +1381,43 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         public void setRevertedHunger(boolean useOldHunger) {
             craftPlayer.getHandle().revertedHunger = useOldHunger;
         }
+
+        private static final int BAR_DRAGON_ENTITY_ID = 12041212;
+        @Override
+        public void setBarMessageWithValue(String message, float value) {
+            PacketPlayOutSpawnEntityLiving entitySpawnPacket = new PacketPlayOutSpawnEntityLiving();
+            entitySpawnPacket.a = BAR_DRAGON_ENTITY_ID;
+            entitySpawnPacket.b = 63;
+            entitySpawnPacket.c = (int)getHandle().locX * 32;
+            entitySpawnPacket.d = -200 * 32;
+            entitySpawnPacket.e = (int)getHandle().locZ * 32;
+            entitySpawnPacket.l = getEnderDragonDataWatcher(message , value);
+            getHandle().playerConnection.sendPacket(entitySpawnPacket);
+        }
+
+        @Override
+        public void destroyBarMessage() {
+            PacketPlayOutEntityDestroy entityDestroyPacket = new PacketPlayOutEntityDestroy(BAR_DRAGON_ENTITY_ID);
+            getHandle().playerConnection.sendPacket(entityDestroyPacket);
+        }
+
+        @Override
+        public void updateMessageWithValue(String message, float value) {
+            PacketPlayOutEntityMetadata packetEntityMeta = new PacketPlayOutEntityMetadata(BAR_DRAGON_ENTITY_ID, getEnderDragonDataWatcher(message, value) , true);
+            getHandle().playerConnection.sendPacket(packetEntityMeta);
+        }
+
+        private DataWatcher getEnderDragonDataWatcher(String text, float value){
+            //Why? using wrapper class CLARIFY which object we are using and wont use wrong object due to ambiguous Object Parameter
+            DataWatcher watcher = new DataWatcher(null);
+            watcher.a(0, Byte.valueOf((byte) 32));
+            watcher.a(6, Float.valueOf(Math.max(1, value)));
+            watcher.a(7, Integer.valueOf(0));
+            watcher.a(8, Byte.valueOf((byte) 0));
+            watcher.a(10, text);
+            watcher.a(11, Byte.valueOf((byte) 1));
+            return watcher;
+        }
     };
 
     public Kyoka kyoka() {
-- 
1.9.4.msysgit.1

