From 30a813a5acdf1f2dacb2bd093476f9cc00c96243 Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Sat, 30 Aug 2014 14:38:24 +0900
Subject: [PATCH] API: Ability to send its dead body to another player

diff --git a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
index ccd4cec..1dc6a42 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
@@ -11,7 +11,7 @@ import java.io.IOException; // CraftBukkit
 
 public class PacketPlayOutNamedEntitySpawn extends Packet {
 
-    private int a;
+    public int a;//Kyoka
     private GameProfile b;
     private int c;
     private int d;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ffa88f4..85116f3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -51,6 +51,7 @@ import org.bukkit.map.MapView;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.StandardMessenger;
+import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.scoreboard.Scoreboard;
 import org.bukkit.util.Vector;
 
@@ -1314,6 +1315,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     // Kyoka start
     private final CraftPlayer craftPlayer = this;
+    private static int entityId = 12041213;
+    private static DataWatcher DEATH_DATA_WATCHER = getDeathDataWatcher();
+
     private final Kyoka kyoka = new Kyoka() {
         @Override
         public String getLocale() {
@@ -1398,6 +1402,26 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             getHandle().playerConnection.sendPacket(packetEntityMeta);
         }
 
+        @Override
+        public void dropDeadBodyCopy(Player viewer , Plugin plugin) {
+            int id = entityId++;
+            final EntityPlayer viewingEntityPlayer = ((CraftPlayer)viewer).getHandle();
+            final PacketPlayOutEntityDestroy packetEntityDestroy = new PacketPlayOutEntityDestroy(id);
+            PacketPlayOutEntityMetadata packetEntityMetaData = new PacketPlayOutEntityMetadata(id , DEATH_DATA_WATCHER , true);
+            PacketPlayOutNamedEntitySpawn packetNamedEntitySpawn = new PacketPlayOutNamedEntitySpawn(getHandle());
+            packetNamedEntitySpawn.a = id;
+            viewingEntityPlayer.playerConnection.sendPacket(packetNamedEntitySpawn);
+            viewingEntityPlayer.playerConnection.sendPacket(packetEntityMetaData);
+            new BukkitRunnable(){
+                @Override
+                public void run() {
+                    if(!viewingEntityPlayer.playerConnection.isDisconnected()){
+                        viewingEntityPlayer.playerConnection.sendPacket(packetEntityDestroy);
+                    }
+                }
+            }.runTaskLater(plugin , 100);
+        }
+
         private DataWatcher getEnderDragonDataWatcher(String text, float value){
             //Why? using wrapper class CLARIFY which object we are using and wont use wrong object due to ambiguous Object Parameter
             DataWatcher watcher = new DataWatcher(null);
@@ -1411,6 +1435,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
     };
 
+    private static DataWatcher getDeathDataWatcher() {
+        DataWatcher watcher = new DataWatcher(null);
+        watcher.a(0, Byte.valueOf((byte) 0));
+        watcher.a(1, Short.valueOf((short) 300));
+        watcher.a(6, Float.valueOf(0.0F));
+        watcher.a(7, Integer.valueOf(0));
+        watcher.a(8, Byte.valueOf((byte) 1));
+        watcher.a(9, Byte.valueOf((byte) 0));
+        watcher.a(16, Byte.valueOf((byte) 0));
+        watcher.a(17, Float.valueOf(0.0F));
+        watcher.a(18, Integer.valueOf(0));
+        return watcher;
+    }
+
     public Kyoka kyoka() {
         return kyoka;
     }
-- 
1.9.4.msysgit.1

