From 318dcd5058df3589fcea935bcbab2bd5438ae48e Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Sat, 27 Sep 2014 13:19:13 +0900
Subject: [PATCH] FEAT: Significantly harder to ender pearl glitch


diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index 650005f..8b0ecce 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -109,11 +109,11 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
         }
 
         Vec3D vec3d = Vec3D.a(this.locX, this.locY, this.locZ);
-        Vec3D vec3d1 = Vec3D.a(this.locX + this.motX, this.locY + this.motY, this.locZ + this.motZ);
+        Vec3D vec3d1 = Vec3D.a(this.locX + (this.motX * 1.5) + (this.motX < 0 ? - 0.25 : 0.25) , this.locY + ((this.motY * 1.5) + 0.2) + (this.motY < 0 ? - 0.25 : 0.25), this.locZ + (this.motZ * 1.5) + (this.motZ < 0 ? - 0.25 : 0.25));
         MovingObjectPosition movingobjectposition = this.world.a(vec3d, vec3d1);
 
         vec3d = Vec3D.a(this.locX, this.locY, this.locZ);
-        vec3d1 = Vec3D.a(this.locX + this.motX, this.locY + this.motY, this.locZ + this.motZ);
+        vec3d1 = Vec3D.a(this.locX + (this.motX * 1.5) + (this.motX < 0 ? - 0.25 : 0.25) , this.locY + ((this.motY * 1.5) + 0.2) + (this.motY < 0 ? - 0.25 : 0.25), this.locZ + (this.motZ * 1.5) + (this.motZ < 0 ? - 0.25 : 0.25));
         if (movingobjectposition != null) {
             vec3d1 = Vec3D.a(movingobjectposition.pos.a, movingobjectposition.pos.b, movingobjectposition.pos.c);
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index cd78e15..81e4c23 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -33,6 +33,11 @@ import net.minecraft.server.PacketPlayOutSetSlot;
 import net.minecraft.server.Slot;
 import net.minecraft.server.World;
 import net.minecraft.server.WorldServer;
+import net.minecraft.server.Item;
+import net.minecraft.server.MathHelper;
+import net.minecraft.server.Vec3D;
+import net.minecraft.server.MovingObjectPosition;
+import net.minecraft.server.EnumMovingObjectType;
 
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
@@ -222,6 +227,11 @@ public class CraftEventFactory {
         }
 
         PlayerInteractEvent event = new PlayerInteractEvent(player, action, itemInHand, blockClicked, blockFace);
+        if(!who.abilities.canInstantlyBuild && itemstack != null && Item.getId(itemstack.getItem()) == 368){
+            if(willLaunchInBlock(who)){
+                event.setCancelled(true);
+            }
+        }
         craftServer.getPluginManager().callEvent(event);
 
         return event;
@@ -905,4 +915,32 @@ public class CraftEventFactory {
         entityHuman.world.getServer().getPluginManager().callEvent(event);
         return (Cancellable) event;
     }
+
+    private static boolean willLaunchInBlock(EntityHuman entityhuman){
+        double motX = (double) (-MathHelper.sin(entityhuman.yaw / 180.0F * 3.1415927F) * MathHelper.cos(entityhuman.pitch / 180.0F * 3.1415927F) * 0.4F);
+        double motZ = (double) (MathHelper.cos(entityhuman.yaw / 180.0F * 3.1415927F) * MathHelper.cos(entityhuman.pitch / 180.0F * 3.1415927F) * 0.4F);
+        double motY = (double) (-MathHelper.sin((entityhuman.pitch) / 180.0F * 3.1415927F) * 0.4F);
+
+        float f2 = MathHelper.sqrt(motX * motX + motY * motY + motZ * motZ);
+
+        motX /= (double) f2;
+        motY /= (double) f2;
+        motZ /= (double) f2;
+        motX += 0.5 * 0.007499999832361937D;
+        motY += 0.5 * 0.007499999832361937D;
+        motZ += 0.5 * 0.007499999832361937D;
+        motX *= 1.5;
+        motY *= 1.5;
+        motZ *= 1.5;
+
+        double locX = entityhuman.locX - (double) (MathHelper.cos(entityhuman.yaw / 180.0F * 3.1415927F) * 0.16F);
+        double locY = entityhuman.locY + entityhuman.getHeadHeight() - 0.10000000149011612D;
+        double locZ=  entityhuman.locZ - (double) (MathHelper.sin(entityhuman.yaw / 180.0F * 3.1415927F) * 0.16F);
+
+        Vec3D vec3d = Vec3D.a(locX ,locY, locZ);
+        Vec3D vec3d1 = Vec3D.a(locX + (motX * 1.5) + (motX < 0 ? - 0.25 : 0.25) ,locY + ((motY * 1.5) + 0.2) + (motY < 0 ? - 0.25 : 0.25), locZ + (motZ * 1.5) + (motZ < 0 ? - 0.25 : 0.25));
+        MovingObjectPosition movingobjectposition = entityhuman.world.a(vec3d, vec3d1);
+
+        return movingobjectposition  != null && movingobjectposition.type == EnumMovingObjectType.BLOCK;
+    }
 }
-- 
1.9.4.msysgit.2

