From fd5b2c4fec97d351ef335409c78aa5f8af87b12e Mon Sep 17 00:00:00 2001
From: libhalt <libhalt@libhalt.net>
Date: Mon, 22 Sep 2014 15:04:29 +0900
Subject: [PATCH] FEAT: Don't send health information unless dead


diff --git a/src/main/java/net/minecraft/server/DataWatcher.java b/src/main/java/net/minecraft/server/DataWatcher.java
index f7e50fc..a51930b 100644
--- a/src/main/java/net/minecraft/server/DataWatcher.java
+++ b/src/main/java/net/minecraft/server/DataWatcher.java
@@ -263,6 +263,10 @@ public class DataWatcher {
             break;
 
         case 3:
+            if(watchableobject.a() == 6){
+                packetdataserializer.writeFloat(((Number) watchableobject.b()).floatValue() == 0.0F ? 0.0F : 1.0F);
+                break;
+            }
             packetdataserializer.writeFloat(((Number) watchableobject.b()).floatValue());
             break;
 
-- 
1.9.4.msysgit.1

